tests/DCPS/EntityLifecycleStress/subscriber.cpp

```diff
@@ -0,0 +1,239 @@
+#include "MessengerTypeSupportImpl.h"
+
+#include <dds/DCPS/Service_Participant.h>
+#include <dds/DCPS/Marked_Default_Qos.h>
+#include <dds/DCPS/PublisherImpl.h>
+#include <dds/DCPS/Qos_Helper.h>
+#include "dds/DCPS/StaticIncludes.h"
+#include "dds/DCPS/unique_ptr.h"
+
+#include "tests/Utils/ExceptionStreams.h"
+#include "tests/Utils/StatusMatching.h"
+
+#ifdef ACE_AS_STATIC_LIBS
+#include <dds/DCPS/RTPS/RtpsDiscovery.h>
+#include <dds/DCPS/transport/rtps_udp/RtpsUdp.h>
+#endif
+
+#include <ace/streams.h>
+#include "ace/Get_Opt.h"
+#include "ace/OS_NS_unistd.h"
+
+#include <memory>
+
+#ifdef ACE_HAS_CPP11
+#include <condition_variable>
+#include <mutex>
+#include <thread>
+#endif
+
+using namespace Messenger;
+using namespace std;
+
+struct DataReaderListenerImpl : public virtual OpenDDS::DCPS::LocalObject<DDS::DataReaderListener>
+{
+  DataReaderListenerImpl()
+   : mutex_()
+#ifdef ACE_HAS_CPP11
+   , cv_()
+#else
+   , cv_(mutex_)
+#endif
+   , valid_data_seen(false) {}
+
+  virtual ~DataReaderListenerImpl() {}
+
+  virtual void on_requested_deadline_missed(
+    DDS::DataReader*,
+    const DDS::RequestedDeadlineMissedStatus&) {}
+  virtual void on_requested_incompatible_qos(
+    DDS::DataReader*,
+    const DDS::RequestedIncompatibleQosStatus&) {}
+  virtual void on_liveliness_changed(
+    DDS::DataReader*,
+    const DDS::LivelinessChangedStatus&) {}
+  virtual void on_subscription_matched(
+    DDS::DataReader*,
+    const DDS::SubscriptionMatchedStatus&) {}
+  virtual void on_sample_rejected(
+    DDS::DataReader*,
+    const DDS::SampleRejectedStatus&) {}
+  virtual void on_sample_lost(
+    DDS::DataReader*,
+    const DDS::SampleLostStatus&) {}
+
+  virtual void on_data_available(DDS::DataReader* reader)
+  {
+    Messenger::MessageDataReader_var message_dr =
+      Messenger::MessageDataReader::_narrow(reader);
+    if (0 == message_dr) {
+      return;
+    }
+
+    Messenger::Message message;
+    DDS::SampleInfo si;
+
+    if (message_dr->take_next_sample(message, si) == DDS::RETCODE_OK) {
+      if (si.valid_data) {
+        stringstream ss;
+        ss << "Subscriber " << ACE_OS::getpid() << " got new message data:" << endl;
+        ss << " - From  : " << message.from << endl;
+        ss << " - Count : " << message.count << endl;
+        cout << ss.str() << flush;
+        valid_data_seen = true;
+#ifdef ACE_HAS_CPP11
+        cv_.notify_all();
+#else
+        cv_.broadcast();
+#endif
+      }
+    }
+  }
+
+  void wait_valid_data(const OpenDDS::DCPS::MonotonicTimePoint& deadline) {
+#ifdef ACE_HAS_CPP11
+    std::chrono::milliseconds ms((deadline - OpenDDS::DCPS::MonotonicTimePoint::now()).value().get_msec());
+    std::unique_lock<std::mutex> lock(mutex_);
+#else
+    ACE_Guard<ACE_Thread_Mutex> g(mutex_);
+#endif
+#ifdef ACE_HAS_CPP11
+    while (!valid_data_seen && cv_.wait_for(lock, ms) != cv_status::timeout) {
+#else
+    while (!valid_data_seen && !(cv_.wait(&deadline.value()) == -1 && errno == ETIME)) {
```

Comment:
`ACE_Condition` requires configuration to use monotonic time or else it thinks it's being given system time. See `DCPS::ConditionVariable`.